File: fixtures/csharp/type_inference/generics/generics.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class GenericsDemo
{
    public static string PrintPair<T, U>(T a, U b)
    {
        return $"{a}: {b}";
    }

    public static List<T> WrapInList<T>(T item)
    {
        return new List<T> { item };
    }

    public static T Identity<T>(T x)
    {
        return x;
    }

    public static Dictionary<K, V> ToMap<K, V>(K key, V value) where K : notnull
    {
        return new Dictionary<K, V> { { key, value } };
    }

    public class Container<T>
    {
        public T Value { get; }

        public Container(T value)
        {
            Value = value;
        }

        public Container<T> Duplicate()
        {
            return new Container<T>(Value);
        }
    }

    public class Pair<A, B>
    {
        public A First { get; }
        public B Second { get; }

        public Pair(A first, B second)
        {
            First = first;
            Second = second;
        }

        public Pair<B, A> Swap()
        {
            return new Pair<B, A>(Second, First);
        }
    }

    public static T Max<T>(T a, T b) where T : IComparable<T>
    {
        return a.CompareTo(b) >= 0 ? a : b;
    }

    public static void Main()
    {
        string pair = PrintPair(42, "hello");
        List<string> list = WrapInList("item");
        string id = Identity("test");
        var map = ToMap("key", 100);
        var container = new Container<string>("value");
        var dup = container.Duplicate();
        var p = new Pair<int, string>(1, "one");
        var swapped = p.Swap();
        int biggest = Max(10, 20);
    }
}

File: fixtures/java/type_inference/generics/generics.java
import java.util.*;

public class GenericsDemo {

    public static <T, U> String printPair(T a, U b) {
        return a.toString() + ": " + b.toString();
    }

    public static <T> List<T> wrapInList(T item) {
        List<T> list = new ArrayList<>();
        list.add(item);
        return list;
    }

    public static <T> T identity(T x) {
        return x;
    }

    public static <K, V> Map<K, V> toMap(K key, V value) {
        Map<K, V> map = new HashMap<>();
        map.put(key, value);
        return map;
    }

    static class Container<T> {
        private T value;

        public Container(T value) {
            this.value = value;
        }

        public T getValue() {
            return value;
        }

        public Container<T> duplicate() {
            return new Container<>(value);
        }
    }

    static class Pair<A, B> {
        private A first;
        private B second;

        public Pair(A first, B second) {
            this.first = first;
            this.second = second;
        }

        public A getFirst() { return first; }
        public B getSecond() { return second; }

        public Pair<B, A> swap() {
            return new Pair<>(second, first);
        }
    }

    public static <T extends Comparable<T>> T max(T a, T b) {
        return a.compareTo(b) >= 0 ? a : b;
    }

    public static void main(String[] args) {
        String pair = printPair(42, "hello");
        List<String> list = wrapInList("item");
        String id = identity("test");
        Map<String, Integer> map = toMap("key", 100);
        Container<String> container = new Container<>("value");
        Container<String> dup = container.duplicate();
        Pair<Integer, String> p = new Pair<>(1, "one");
        Pair<String, Integer> swapped = p.swap();
        Integer biggest = max(10, 20);
    }
}
